File: packages/contracts/workspace-image-release.ts
export interface WorkspaceImageRelease {
  readonly version: string;
  readonly image: string;
}

export interface WorkspaceImageReleaseCatalog {
  readonly schemaVersion: 1;
  readonly releases: readonly WorkspaceImageRelease[];
}

const digestPattern = /^sha256:[a-f0-9]{64}$/;
const versionPattern = /^[A-Za-z0-9][A-Za-z0-9._+-]{0,63}$/;

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: readonly string[]): boolean {
  const keys = Object.keys(value).sort();
  return keys.length === expected.length && keys.every((key, index) => key === [...expected].sort()[index]);
}

export function validWorkspaceImageReference(value: string): boolean {
  const normalized = value.trim();
  const separator = normalized.indexOf("@");
  if (separator <= 0 || normalized.indexOf("@", separator + 1) !== -1) return false;
  return digestPattern.test(normalized.slice(separator + 1));
}

export function decodeWorkspaceImageReleaseCatalog(raw: string, installedImage: string): WorkspaceImageReleaseCatalog | null {
  const installed = installedImage.trim();
  if (!validWorkspaceImageReference(installed)) return null;
  if (!raw.trim()) {
    return { schemaVersion: 1, releases: [{ version: "installed", image: installed }] };
  }

  let decoded: unknown;
  try {
    decoded = JSON.parse(raw);
  } catch {
    return null;
  }
  if (!isRecord(decoded) || !hasExactKeys(decoded, ["releases", "schemaVersion"]) || decoded.schemaVersion !== 1 ||
    !Array.isArray(decoded.releases) || decoded.releases.length < 1 || decoded.releases.length > 32) return null;

  const releases: WorkspaceImageRelease[] = [];
  const versions = new Set<string>();
  const images = new Set<string>();
  for (const item of decoded.releases) {
    if (!isRecord(item) || !hasExactKeys(item, ["image", "version"]) || typeof item.version !== "string" || typeof item.image !== "string") return null;
    const version = item.version.trim();
    const image = item.image.trim();
    if (!versionPattern.test(version) || !validWorkspaceImageReference(image) || versions.has(version) || images.has(image)) return null;
    versions.add(version);
    images.add(image);
    releases.push({ version, image });
  }
  return images.has(installed) ? { schemaVersion: 1, releases } : null;
}

File: packages/contracts/go/workspace_image_release.go
package contracts

import (
	"encoding/json"
	"errors"
	"io"
	"regexp"
	"strings"
)

const WorkspaceImageReleasesEnv = "OPL_WORKSPACE_IMAGE_RELEASES_JSON"

var (
	workspaceImageDigestPattern  = regexp.MustCompile(`^sha256:[a-f0-9]{64}$`)
	workspaceImageVersionPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._+-]{0,63}$`)
)

type WorkspaceImageRelease struct {
	Version string `json:"version"`
	Image   string `json:"image"`
}

type WorkspaceImageReleaseCatalog struct {
	SchemaVersion int                     `json:"schemaVersion"`
	Releases      []WorkspaceImageRelease `json:"releases"`
}

type WorkspaceImageReleaseActivationRequest struct {
	ReleaseVersion   string `json:"releaseVersion"`
	ExpectedRevision int    `json:"expectedRevision"`
	Reason           string `json:"reason"`
}

type WorkspaceRuntimeImageReplacementPreview struct {
	WorkspaceID        string `json:"workspaceId"`
	WorkspaceStatus    string `json:"workspaceStatus"`
	RuntimeID          string `json:"runtimeId"`
	RuntimeStatus      string `json:"runtimeStatus"`
	CurrentImageDigest string `json:"currentImageDigest"`
	TargetImageDigest  string `json:"targetImageDigest"`
	CanReplace         bool   `json:"canReplace"`
}

type WorkspaceComputeRuntimeBindingStatus string

const (
	WorkspaceComputeRuntimeBindingMatched    WorkspaceComputeRuntimeBindingStatus = "matched"
	WorkspaceComputeRuntimeBindingMismatched WorkspaceComputeRuntimeBindingStatus = "mismatched"
)

type WorkspaceComputeRuntimeBinding struct {
	Status WorkspaceComputeRuntimeBindingStatus `json:"status"`
}

func ValidWorkspaceImageReference(value string) bool {
	value = strings.TrimSpace(value)
	repository, digest, ok := strings.Cut(value, "@")
	return ok && repository != "" && !strings.Contains(repository, "@") && workspaceImageDigestPattern.MatchString(digest)
}

func WorkspaceImageDigest(value string) string {
	_, digest, ok := strings.Cut(strings.TrimSpace(value), "@")
	if !ok || !workspaceImageDigestPattern.MatchString(digest) {
		return ""
	}
	return digest
}

func DecodeWorkspaceImageReleaseCatalog(raw, installedImage string) (WorkspaceImageReleaseCatalog, error) {
	installedImage = strings.TrimSpace(installedImage)
	if !ValidWorkspaceImageReference(installedImage) {
		return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_installed_image_invalid")
	}
	if strings.TrimSpace(raw) == "" {
		return WorkspaceImageReleaseCatalog{
			SchemaVersion: 1,
			Releases:      []WorkspaceImageRelease{{Version: "installed", Image: installedImage}},
		}, nil
	}

	var catalog WorkspaceImageReleaseCatalog
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&catalog); err != nil {
		return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
	}
	if catalog.SchemaVersion != 1 || len(catalog.Releases) == 0 || len(catalog.Releases) > 32 {
		return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
	}
	versions, images := map[string]struct{}{}, map[string]struct{}{}
	installedFound := false
	for index := range catalog.Releases {
		release := &catalog.Releases[index]
		release.Version = strings.TrimSpace(release.Version)
		release.Image = strings.TrimSpace(release.Image)
		if !workspaceImageVersionPattern.MatchString(release.Version) || !ValidWorkspaceImageReference(release.Image) {
			return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
		}
		if _, exists := versions[release.Version]; exists {
			return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
		}
		if _, exists := images[release.Image]; exists {
			return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_catalog_invalid")
		}
		versions[release.Version], images[release.Image] = struct{}{}, struct{}{}
		installedFound = installedFound || release.Image == installedImage
	}
	if !installedFound {
		return WorkspaceImageReleaseCatalog{}, errors.New("workspace_image_release_installed_image_missing")
	}
	return catalog, nil
}

func (catalog WorkspaceImageReleaseCatalog) ReleaseForImage(image string) (WorkspaceImageRelease, bool) {
	for _, release := range catalog.Releases {
		if release.Image == image {
			return release, true
		}
	}
	return WorkspaceImageRelease{}, false
}

func (catalog WorkspaceImageReleaseCatalog) ContainsImage(image string) bool {
	_, found := catalog.ReleaseForImage(image)
	return found
}
